nixos/tests/wireguard/generated.nix

```diff
@@ -0,0 +1,57 @@
+import ../make-test.nix ({ pkgs, ...} : {
+  name = "wireguard-generated";
+  meta = with pkgs.stdenv.lib.maintainers; {
+    maintainers = [ ma27 ];
+  };
+
+  nodes = {
+    peer1 = {
+      networking.firewall.allowedUDPPorts = [ 12345 ];
+      networking.wireguard.interfaces.wg0 = {
+        ips = [ "10.10.10.1/24" ];
+        listenPort = 12345;
+        privateKeyFile = "/etc/wireguard/private";
+        generatePrivateKeyFile = true;
+
+      };
+    };
+
+    peer2 = {
+      networking.firewall.allowedUDPPorts = [ 12345 ];
+      networking.wireguard.interfaces.wg0 = {
+        ips = [ "10.10.10.2/24" ];
+        listenPort = 12345;
+        privateKeyFile = "/etc/wireguard/private";
+        generatePrivateKeyFile = true;
+      };
+    };
+  };
+
+  testScript = ''
+    startAll;
+
+    $peer1->waitForUnit("wireguard-wg0.service");
+    $peer2->waitForUnit("wireguard-wg0.service");
+
+    my ($retcode, $peer1pubkey) = $peer1->execute("wg pubkey < /etc/wireguard/private");
+    $peer1pubkey =~ s/\s+$//;
+    if ($retcode != 0) {
+      die "Could not read public key from peer1";
+    }
+
+    my ($retcode, $peer2pubkey) = $peer2->execute("wg pubkey < /etc/wireguard/private");
+    $peer2pubkey =~ s/\s+$//;
+    if ($retcode != 0) {
+      die "Could not read public key from peer2";
+    }
+
+    $peer1->succeed("wg set wg0 peer $peer2pubkey allowed-ips 10.10.10.2/32 endpoint 192.168.1.2:12345 persistent-keepalive 1");
```

Comment:
I'm probably missing something, but how would you exchange pubkeys e.g. when deploying two wireguard peers in a nixops network?

Reply:
Left As An Exercise To The Reader.

Frankly I don't know :/,  beyond some (probably somewhat manual) "intake" process during provisioning. It may be that this isn't an option most people would choose to use with Nixops (though I will).

Reply:
(I'm thinking I might integrate with Vault or something, but until then plan to just copy keys around by hand)

Reply:
> Frankly I don't know :/, beyond some (probably somewhat manual) "intake" process during provisioning. It may be that this isn't an option most people would choose to use with Nixops (though I will).

Yeah this is more or less what I'm doing most of the time when using nixops to deploy stuff :)

> (I'm thinking I might integrate with Vault or something, but until then plan to just copy keys around by hand)

Some time ago I played around with Ansible+Vault and it actually felt quite nice to use (especially compared to how I do secret management with nixops at the moment). Unfortunately I'm not that experienced with this tooling, hence I don't consider myself sufficiently qualified to implement such a security-sensitive feature for a deployment tool.

But if you start implementing such a feature, feel free to ping me, I'd love to help out! :)